[Compiling mono problem] fatal error: mono/metadata/qcall-def.h: No such file or directory

Hey mono devs,

This is a compile problem I just had on linux (source tarball of mono).

The URL to the **source** **archive** I used was:

https://download.mono-project.com/sources/mono/nightly/mono-6.13.0.643.tar.xz


Error:


-------




In file included from ../../mono/metadata/w32socket.h:16,
                 from w32socket.c:19:
w32socket.c: In function ‘create_object_handle_from_sockaddr’:
../../mono/metadata/object-internals.h:252:8: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
  252 |   *__p = (value); \
      |   ~~~~~^~~~~~~~~
../../mono/metadata/handle.h:454:4: note: in expansion of macro ‘mono_array_set_internal’
  454 |    mono_array_set_internal (MONO_HANDLE_RAW (MONO_HANDLE_UNSUPPRESS (HANDLE)), TYPE, __idx, __val); \
      |    ^~~~~~~~~~~~~~~~~~~~~~~
w32socket.c:978:5: note: in expansion of macro ‘MONO_HANDLE_ARRAY_SETVAL’
  978 |     MONO_HANDLE_ARRAY_SETVAL (data, guint8, 8 + i, 0);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../../mono/metadata/w32socket.h:16,
                 from w32socket.c:19:
../../mono/metadata/object-internals.h:175:22: note: at offset 0 to object ‘vector’ with size 0 declared here
  175 |  mono_64bitaligned_t vector [MONO_ZERO_LEN_ARRAY];
      |                      ^~~~~~
  CC       libmonoruntime_la-w32process.lo
  CC       libmonoruntime_la-profiler.lo
  CC       libmonoruntime_la-rand.lo
  CC       libmonoruntime_la-remoting.lo
  CC       libmonoruntime_la-runtime.lo
  CC       libmonoruntime_la-mono-security.lo
  CC       libmonoruntime_la-security-core-clr.lo
  CC       libmonoruntime_la-security-manager.lo
  CC       libmonoruntime_la-string-icalls.lo
  CC       libmonoruntime_la-sysmath.lo
  CC       libmonoruntime_la-threads.lo
threads.c: In function ‘mono_threads_summarize_execute_internal’:
threads.c:6592:6: warning: ‘current_idx’ may be used uninitialized in this function [-Wmaybe-uninitialized]
 6592 |  int current_idx;
      |      ^~~~~~~~~~~
  CC       libmonoruntime_la-threadpool.lo
  CC       libmonoruntime_la-threadpool-io.lo
  CC       libmonoruntime_la-verify.lo
  CC       libmonoruntime_la-dynamic-stream.lo
  CC       libmonoruntime_la-file-mmap-posix.lo
  CC       libmonoruntime_la-file-mmap-windows.lo
  CC       libmonoruntime_la-metadata-cross-helpers.lo
  CC       libmonoruntime_la-seq-points-data.lo
  CC       libmonoruntime_la-handle.lo
  CC       libmonoruntime_la-w32handle-namespace.lo
  CC       libmonoruntime_la-w32handle.lo
  CC       libmonoruntime_la-reflection.lo
  CC       libmonoruntime_la-dynamic-image.lo
  CC       libmonoruntime_la-sre.lo
  CC       libmonoruntime_la-sre-encode.lo
  CC       libmonoruntime_la-sre-save.lo
  CC       libmonoruntime_la-custom-attrs.lo
  CC       libmonoruntime_la-fdhandle.lo
  CC       libmonoruntime_la-callspec.lo
  CC       libmonoruntime_la-abi.lo
  CC       libmonoruntime_la-native-library.lo
  CC       libmonoruntime_la-native-library-qcall.lo
native-library-qcall.c:31:14: fatal error: mono/metadata/qcall-def.h: No such file or directory
   31 |     #include "mono/metadata/qcall-def.h"
      |              ^~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[3]: *** [Makefile:2688: libmonoruntime_la-native-library-qcall.lo] Error 1
make[3]: Leaving directory '/home/Temp/rbt/mono-6.13.0.643/mono/metadata'
make[2]: *** [Makefile:542: all-recursive] Error 1
make[2]: Leaving directory '/home/Temp/rbt/mono-6.13.0.643/mono'
make[1]: *** [Makefile:604: all-recursive] Error 1
make[1]: Leaving directory '/home/Temp/rbt/mono-6.13.0.643'
make: *** [Makefile:532: all] Error 2



-------




My system:

  Operating system                GNU/Linux
  Operating system bit type       x86_64 (64 bit)
  CPU Model                       AMD A8-7600 Radeon R7, 10 Compute Cores 4C+6G, 4 cores
  RAM                             14954 MB RAM
  cflags in use                   -O2 -fPIC -fno-strict-overflow -Wno-error
  gcc                             10.1.0
  glibc                           2.23  
  linux kernel                    4.4.14
  binutils                        2.34

(Note that mono compiled on the same machine with virtually the same setup, some weeks
just fine so I assume something may have changed internally. I did not pass any special
configure option, just ./configure --prefix=/usr

Sorry if this bug was already reported).

--- a/mono/metadata/Makefile.am
+++ b/mono/metadata/Makefile.am
@@ -418,7 +418,8 @@ common_sources = \
 	abi.c	\
 	native-library.c \
 	native-library.h \
-	native-library-qcall.c
+	native-library-qcall.c \
+	qcall-def.h 
 
 # These source files have compile time dependencies on GC code
 gc_dependent_sources = \

